src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1182 @@
+                        break;
+                    case State.ResponseHeaders:
+                        _response.Headers.TryAddWithoutValidation(descriptor.HeaderType.HasFlag(HttpHeaderType.Request) ? descriptor.AsCustomHeader() : descriptor, headerValue);
+                        break;
+                    case State.TrailingHeaders:
+                        _trailingHeaders.Add((descriptor.HeaderType.HasFlag(HttpHeaderType.Request) ? descriptor.AsCustomHeader() : descriptor, headerValue));
+                        break;
+                    default:
+                        Debug.Fail($"Unexpected {nameof(Http3RequestStream)}.{nameof(_headerState)} '{_headerState}'.");
+                        break;
+                }
+            }
+        }
+
+        void IHttpHeadersHandler.OnHeadersComplete(bool endStream)
+        {
+            throw new NotImplementedException();
+        }
+
+        private async ValueTask SkipUnknownPayloadAsync(Http3FrameType frameType, long payloadLength, CancellationToken cancellationToken)
+        {
+            if (payloadLength > MaximumUnknownFramePayloadLength)
+            {
+                throw new HttpRequestException($"Frame type {frameType} with payload of length {payloadLength} exceeds maximum unknown frame payload length of of {MaximumUnknownFramePayloadLength}.");
+            }
+
+            while (payloadLength != 0)
+            {
+                if (_buffer.ActiveLength == 0)
+                {
+                    _buffer.EnsureAvailableSpace(1);
+                    int bytesRead = await _stream.ReadAsync(_buffer.AvailableMemory, cancellationToken).ConfigureAwait(false);
+
+                    if (bytesRead != 0)
+                    {
+                        _buffer.Commit(bytesRead);
+                    }
+                    else
+                    {
+                        // Our buffer has partial frame data in it but not enough to complete the read: bail out.
+                        throw new HttpRequestException("Partial frame received; unexpected end of stream.");
+                    }
+                }
+
+                long readLength = Math.Min(payloadLength, _buffer.ActiveLength);
+                _buffer.Discard((int)readLength);
+                payloadLength -= readLength;
+            }
+        }
+
+        private int ReadResponseContent(HttpResponseMessage response, Span<byte> buffer)
+        {
+            // This should only be called by our read stream, and our local _response should be nulled out by now.
+            Debug.Assert(_response == null);
+
+            try
+            {
+                int totalBytesRead = 0;
+
+                while (buffer.Length != 0)
+                {
+                    // Sync over async here -- QUIC implementation does it per-I/O already; this is at least more coarse-grained.
+                    if (_dataPayloadRemaining <= 0 && !ReadNextDataFrameAsync(response, CancellationToken.None).GetAwaiter().GetResult())
```

Comment:
This will probably need to change eventually to no do sync over async.

Reply:
Yea, I had same thought -- leaving as a TODO until we firm up the behavior of QuicStream.